iModelCore/ECDb/ECDb/SchemaWriter.cpp

```diff
@@ -3262,22 +3262,31 @@ BentleyStatus SchemaWriter::DeleteCustomAttributeClass(Context& ctx, ECCustomAtt
 //---------------------------------------------------------------------------------------
 // @bsimethod
 //+---------------+---------------+---------------+---------------+---------------+------
-BentleyStatus SchemaWriter::DeleteClass(Context& ctx, ClassChange& classChange, ECClassCR deletedClass)
+BentleyStatus SchemaWriter::DeleteClass(Context& ctx, ClassChange& classChange, ECClassCR deletedClass, bool isDynamicSchema)
     {
+    auto isMajorSchemaVersionChange = ctx.IsMajorSchemaVersionChange(deletedClass.GetSchema().GetId());
+
     if (!ctx.AreMajorSchemaVersionChangesAllowed() || !ctx.IsMajorSchemaVersionChange(deletedClass.GetSchema().GetId()))
         {
-        if (ctx.IgnoreIllegalDeletionsAndModifications())
+        if (ctx.AreMajorSchemaVersionChangesAllowedForDynamicSchemas() && isMajorSchemaVersionChange && isDynamicSchema)
             {
-            LOG.infov("Ignoring upgrade error:  ECSchema Upgrade failed. ECSchema %s: Cannot delete ECClass '%s'. This is a major ECSchema change. Either major schema version changes are disabled "
-                                 "or the 'Read' version number of the ECSchema was not incremented.",
-                                 deletedClass.GetSchema().GetFullSchemaName().c_str(), deletedClass.GetName().c_str());
-            return SUCCESS;
+            LOG.infov("Major version change detected for dynamic schema. Deleting ECClass '%s.%s.'", deletedClass.GetSchema().GetFullSchemaName().c_str(), deletedClass.GetName().c_str());
             }
+        else
+            {
+            if (ctx.IgnoreIllegalDeletionsAndModifications())
+                {
+                LOG.infov("Ignoring upgrade error:  ECSchema Upgrade failed. ECSchema %s: Cannot delete ECClass '%s'. This is a major ECSchema change. Either major schema version changes are disabled "
+                    "or the 'Read' version number of the ECSchema was not incremented.",
+                    deletedClass.GetSchema().GetFullSchemaName().c_str(), deletedClass.GetName().c_str());
+                return SUCCESS;
+                }
 
-        ctx.Issues().ReportV(IssueSeverity::Error, IssueCategory::BusinessProperties, IssueType::ECDbIssue, "ECSchema Upgrade failed. ECSchema %s: Cannot delete ECClass '%s'. This is a major ECSchema change. Either major schema version changes are disabled "
+            ctx.Issues().ReportV(IssueSeverity::Error, IssueCategory::BusinessProperties, IssueType::ECDbIssue, "ECSchema Upgrade failed. ECSchema %s: Cannot delete ECClass '%s'. This is a major ECSchema change. Either major schema version changes are disabled "
                          "or the 'Read' version number of the ECSchema was not incremented.",
                                   deletedClass.GetSchema().GetFullSchemaName().c_str(), deletedClass.GetName().c_str());
-        return ERROR;
+            return ERROR;
+            }

```

Comment:
Overall I think these changes are making the logic more complex than necessary.  I suggest something like:

```cpp
bool changeIsAllowed = // combine all parameters except IgnoreIllegalDeletionsAndModifications (because it sucks and we want to get rid of it)
if (!changeIsAllowed)
  {
  // Existing gross ignore illegal deletions and modifications code
  }
```

This doesn't give a special log message for major version change allowed due to dynamic schema but maybe there is a better place to put such a log besides down in delete class.  Also if there is no log for a class being deleted I can see adding a log saying a class is being deleted but it should happen no matter what the circumstances are.

Reply:
Or maybe `bool changeNotAllowed` instead of immediately inverting the bool

Reply:
Done !